File: packages/java-client/wrapper/src/main/java/com/consentshield/sdk/ConsentShieldApiException.java
package com.consentshield.sdk;

import java.util.Map;

/**
 * Thrown when the ConsentShield API returns a 4xx response.
 *
 * <p>4xx responses ALWAYS surface — they are never retried, never folded into
 * fail-OPEN handling. The compliance contract requires the caller to see the
 * problem+json detail (RFC 7807) so they can correct the request.
 */
public final class ConsentShieldApiException extends ConsentShieldException {
    private final int status;
    private final String type;
    private final String title;
    private final String detail;
    private final String instance;
    private final String traceId;
    private final Map<String, Object> extensions;

    public ConsentShieldApiException(
            int status,
            String type,
            String title,
            String detail,
            String instance,
            String traceId,
            Map<String, Object> extensions) {
        super(buildMessage(status, title, detail));
        this.status = status;
        this.type = type;
        this.title = title;
        this.detail = detail;
        this.instance = instance;
        this.traceId = traceId;
        this.extensions = extensions == null ? Map.of() : Map.copyOf(extensions);
    }

    public int getStatus() {
        return status;
    }

    public String getType() {
        return type;
    }

    public String getTitle() {
        return title;
    }

    public String getDetail() {
        return detail;
    }

    public String getInstance() {
        return instance;
    }

    @Override
    public String getTraceId() {
        return traceId;
    }

    public Map<String, Object> getExtensions() {
        return extensions;
    }

    private static String buildMessage(int status, String title, String detail) {
        StringBuilder sb = new StringBuilder("ConsentShield API error: status=").append(status);
        if (title != null && !title.isEmpty()) sb.append(" title=").append(title);
        if (detail != null && !detail.isEmpty()) sb.append(" detail=").append(detail);
        return sb.toString();
    }
}

File: packages/php-client/wrapper/src/Exception/ConsentShieldApiException.php
<?php declare(strict_types=1);

namespace ConsentShield\Sdk\Exception;

/**
 * Thrown when the ConsentShield API returns a 4xx response.
 * 4xx responses ALWAYS surface — never retried, never folded into fail-OPEN.
 */
final class ConsentShieldApiException extends ConsentShieldException
{
    /**
     * @param array<string, mixed> $extensions
     */
    public function __construct(
        private readonly int $status,
        private readonly ?string $type,
        private readonly ?string $title,
        private readonly ?string $detail,
        private readonly ?string $instance,
        ?string $traceId,
        private readonly array $extensions = [],
    ) {
        $msg = "ConsentShield API error: status={$status}"
            . ($title !== null && $title !== '' ? " title={$title}" : '')
            . ($detail !== null && $detail !== '' ? " detail={$detail}" : '');
        parent::__construct($msg, $traceId);
    }

    public function getStatus(): int { return $this->status; }
    public function getType(): ?string { return $this->type; }
    public function getTitle(): ?string { return $this->title; }
    public function getDetail(): ?string { return $this->detail; }
    public function getInstance(): ?string { return $this->instance; }
    /** @return array<string, mixed> */
    public function getExtensions(): array { return $this->extensions; }
}
